common/src/main/java/com/flipkart/varadhi/common/EntityReadCache.java

```diff
@@ -0,0 +1,166 @@
+import java.util.stream.Collectors;
+
+/**
+ * A thread-safe, generic cache for entities with event handling capabilities.
+ * <p>
+ * This class provides concurrent implementation for caching entities with the following features:
+ * <ul>
+ *   <li>Thread-safe operations using {@link ConcurrentHashMap}</li>
+ *   <li>Reactive preloading with {@link Future} for asynchronous operations</li>
+ *   <li>Type-safe generic implementation constrained to {@link MetaStoreEntity} types</li>
+ * </ul>
+ * <p>
+ * This class is optimized for concurrent access patterns with a focus on read-heavy workloads,
+ * making it suitable where entity data changes infrequently compared to read operations.
+ *
+ * @param <T> the type of entity managed by this cache, must extend {@link MetaStoreEntity}
+ */
+@Slf4j
+public class EntityReadCache<T extends MetaStoreEntity> implements EntityEventListener<T> {
+
+    /**
+     * Thread-safe storage for entities, indexed by their names.
+     */
+    protected final ConcurrentHashMap<String, T> entities;
+
+    /**
+     * The resource type managed by this provider.
+     */
+    @Getter
+    protected final ResourceType resourceType;
+
+    /**
+     * The strategy for loading all entities.
+     */
+    private final Supplier<List<T>> entityLoader;
+
+    /**
+     * Creates a new entity cache for the specified resource type with the provided loading strategy.
+     *
+     * @param resourceType the type of resource managed by this provider
+     * @param entityLoader the strategy for loading all entities
+     * @throws NullPointerException if resourceType or entityTypeName is null
+     */
+    public EntityReadCache(ResourceType resourceType, Supplier<List<T>> entityLoader) {
+        this.resourceType = Objects.requireNonNull(resourceType, "Resource type cannot be null");
+        this.entityLoader = Objects.requireNonNull(entityLoader, "Entity loader cannot be null");
+        this.entities = new ConcurrentHashMap<>();
+        log.debug("Initialized {} provider", resourceType);
+    }
+
+    /**
+     * Preloads all entities into the in-memory cache.
+     * <p>
+     * This method fetches all entities using the configured entity loader and
+     * populates the cache. It is thread-safe and can be called concurrently.
+     * <p>
+     * The operation is performed asynchronously and returns a {@link Future} that
+     * completes when the preload operation is finished.
+     *
+     * @return a future that completes when the preload operation is finished
+     */
+    public Future<Void> preload() {
+        Promise<Void> promise = Promise.promise();
+
+        try {
+            log.info("Starting to preload {}", resourceType.toString());
+            List<T> entityList = entityLoader.get();
+
+            if (entityList.isEmpty()) {
+                log.info("No {} found to preload", resourceType);
+                promise.complete();
+                return promise.future();
+            }
```

Comment:
why is this code required?

Below when you created the map, there itself you are printing the size, and in case of 0 size, that will be printed as well.